django/db/models/sql/query.py

```diff
@@ -569,6 +569,22 @@ def combine(self, rhs, connector):
         if self.distinct_fields != rhs.distinct_fields:
             raise TypeError('Cannot combine queries with different distinct fields.')
 
+        # If `lhs` and `rhs` shares the same alias prefix, it is possible to have
+        # conflicting alias changes like T4 -> T5 and T5 -> T6 (which ends up as
+        # T4 -> T6) while combining two querysets. To prevent this, change alias
+        # prefix of the `rhs` and update current aliases accordingly, except if the
+        # alias is the base table since it needs to be present in both lhs and rhs.
+        if self.alias_prefix == rhs.alias_prefix:
+            old_alias_prefix = rhs.alias_prefix
+            rhs.bump_prefix(self, change_aliases=False)
+            initial_alias = rhs.get_initial_alias()
+            rhs.change_aliases({
+                alias: alias.replace(old_alias_prefix, rhs.alias_prefix)
+                for alias, join in rhs.alias_map.items()
+                if join.table_alias not in [join.table_name, initial_alias]
+            })
+            rhs.unref_alias(initial_alias)
```

Comment:
I have a feeling that it could be more consistent. We pass `change_aliases=False` to avoid changing aliases, then change them here in a similar way as `bump_prefix()` does :thinking:.

Is it not enough to pass `self.get_initial_alias()` to the `bump_prefix()` and exclude matching aliases from [the final loop](https://github.com/django/django/blob/24b316536a7ee4c54a54f632de1852aecb4038c0/django/db/models/sql/query.py#L923-L926)?

Reply:
I guess it will be more meaningful to implement it after deciding the solution. The reason why we use `change_aliases` outside of the `bump_prefix` is that `bump_prefix` also creates an alias for the tables that had no alias before. To prevent this, I have a suggestion. What about deleting `change_aliases` parameter of `bump_prefix` and instead add a new optional parameter called `aliases` (or `aliases_to_change` to be more verbose, could you please give me your opinion?). If `aliases_to_change` is provided then we can only change the provided aliases, else we can keep the original behavior of the `bump_prefix` method. What do you think?

Reply:
Another thing is that, if we implement the solution above then the `L640`:

`self.subq_aliases |= rhs.subq_aliases`

seems to be redundant as we do this in the `bump_prefix`. However, we should remove the if check in the `L577`:

`if self.alias_prefix == rhs.alias_prefix:`

to make sure that `self.subq_aliases` is updated. The same check is already done in the `bump_prefix`, `L952`. Is it necessary to update `self.subq_aliases` even though the `alias_prefix` of `rhs` and `lhs` are different?

Reply:
@omerfarukabaci

> Is it necessary to update self.subq_aliases even though the alias_prefix of rhs and lhs are different?
> Edit: On second thoughts I think it is necessary to update self.subq_aliases even if the alias_prefix of rhs and lhs are different. Could you please correct me?

I had the same line of thoughts and came to the same conclusion; I think we'll want to keep `self.subq_aliases |= rhs.subq_aliases` around to make sure it's applied even combined queries share the same alias. One way to confirm would be to create a variant of the test added in 6d0cbe42c3d382e5393d4af48185c546bb0ada1f that causes queries to have a different alias prefix

https://github.com/django/django/blob/28f66b2783e52d5be51887c31cd2be907b820e11/tests/queries/tests.py#L2122-L2131

Reply:
@charettes

Thanks for your reply! I think I got your point and I agree. If we compromise on the implementation steps below, I can make the related changes:

1. Add a new test for `subquery_aliases` (similar to the one that you have provided) where `lhs` and `rhs` don't share the same alias.
2. Replace `change_aliases` parameter of `Query.bump_prefix` with an optional `change_map` parameter (if you have better naming suggestions, please let me know). Within the method change aliases using the provided `change_map` if any, else keep the current behaviour.
3. Call `rhs.bump_prefix(change_map=rhs_change_map)`, where `rhs_change_map` is the [same change map above](https://github.com/omerfarukabaci/django/blob/ticket_33319/django/db/models/sql/query.py#L582-L584).
4. Remove the [redundant if check](https://github.com/omerfarukabaci/django/blob/ticket_33319/django/db/models/sql/query.py#L577) since it is already done in the `Query.bump_prefix`.
5. We will keep [this line](https://github.com/omerfarukabaci/django/blob/ticket_33319/django/db/models/sql/query.py#L640) even though this logic is duplicated if `rhs` and `lhs` have the same prefix, since it doesn't have any side-effect and it is not a complex operation.

Reply:
@omerfarukabaci that sounds like a good plan, thanks for drafting it up!

The extra tests for 1. should likely be done in prior separate commit.

> Replace change_aliases parameter of Query.bump_prefix with an optional change_map parameter (if you have better naming suggestions, please let me know). Within the method change aliases using the provided change_map if any, else keep the current behaviour.

An alternative could be to pass an `exclude={self.get_initial_alias()}` kwarg and do `if alias not in exclude` in the `change_aliases` comprehension?

Reply:
@charettes You're welcome!

> Add a new test for subquery_aliases (similar to the one that you have provided) where lhs and rhs don't share the same alias.

Apparently this is already tested in the new test I have added, `test_conflicting_change_map_during_combine`.

In the line `self.assertEqual(qs1 | qs2, qs2 | qs1)`, during the operation of `qs1 | qs2`, `alias_prefix` of `qs2` is bumped, after that during the operation `qs2 | qs1`, they don't share the same `alias_prefix` anymore (they are `U` and `T`), and `subq_aliases` are `{'T'}` for `qs2` and `{'T', 'U'}` for `qs1`. IMHO it should be sufficient, and there is no need to add another test. What do you think?

Reply:
Related changes are made, commits are squashed. I would appreciate your review. 🚀